streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -123,30 +126,40 @@ public StreamTask(final TaskId id,
             partitionQueues.put(partition, queue);
         }
 
-        partitionGroup = new PartitionGroup(partitionQueues, timestampExtractor);
+        partitionGroup = new PartitionGroup(partitionQueues);
 
         // initialize the consumed offset cache
         consumedOffsets = new HashMap<>();
 
-        // create the record recordCollector that maintains the produced offsets
-        this.recordCollector = recordCollector;
+        this.producer = producer;
+        recordCollector = createRecordCollector();
 
         // initialize the topology with its own context
         processorContext = new ProcessorContextImpl(id, this, config, recordCollector, stateMgr, metrics, cache);
         this.time = time;
         log.debug("{} Initializing", logPrefix);
         initializeStateStores();
         stateMgr.registerGlobalStateStores(topology.globalStateStores());
+        if (exactlyOnceEnabled) {
+            producer.initTransactions();
+            producer.beginTransaction();
+        }
         initTopology();
         processorContext.initialized();
     }
 
     /**
-     * re-initialize the task
+     * <pre>
+     * - re-initialize the task
+     * - if (eos) begin new transaction
+     * </pre>
      */
     @Override
     public void resume() {
         log.debug("{} Resuming", logPrefix);
+        if (exactlyOnceEnabled) {
+            producer.beginTransaction();
```

Comment:
It is not possible to check -- and at this point, it should never be the case. Producer would throw an exception if there is an open transaction (and this would indicate a bug in our code -- thus, we should not catch this exception but fail IMHO).